File: python/examples/a2a-tools/main.py
"""Two flavours of Agent2Agent delegation in one ephemeral agent.

  * ``mantyx_a2a`` — public peer MANTYX dials directly (server-resolved).
  * ``define_local_a2a`` — peer the SDK reaches on MANTYX's behalf
    (client-resolved). Pass an ``agent_card_url``; the SDK fetches the
    Agent Card on the first run, ships it inline, and speaks A2A
    ``message/send`` to ``agent_card.url`` whenever MANTYX emits a
    ``local_tool_call`` for this tool. You only supply the URL.

Usage:
    export MANTYX_API_KEY=mtx_live_...
    export MANTYX_WORKSPACE_SLUG=acme-corp
    # Optional — register a real public peer:
    # export BILLING_AGENT_CARD_URL=https://billing.example.com/.well-known/agent-card.json
    # export BILLING_AGENT_TOKEN=...
    # Required — your intranet peer's Agent Card URL:
    export HR_AGENT_CARD_URL=https://hr.intranet/.well-known/agent-card.json
    # export HR_AGENT_TOKEN=...
    python main.py "How do I expense client lunches?"
"""

from __future__ import annotations

import os
import sys

from mantyx import MantyxClient, ToolRef, define_local_a2a, mantyx_a2a


def required_env(name: str) -> str:
    v = os.environ.get(name)
    if not v:
        print(f"Missing required env var {name}", file=sys.stderr)
        sys.exit(1)
    return v


def main() -> None:
    client = MantyxClient(
        api_key=required_env("MANTYX_API_KEY"),
        workspace_slug=required_env("MANTYX_WORKSPACE_SLUG"),
        base_url=os.environ.get("MANTYX_BASE_URL", "https://app.mantyx.io"),
    )

    tools: list[ToolRef] = []

    if os.environ.get("BILLING_AGENT_CARD_URL"):
        token = os.environ.get("BILLING_AGENT_TOKEN")
        tools.append(
            mantyx_a2a(
                name="billing_agent",
                description="Delegate billing questions to the public Acme billing agent.",
                agent_card_url=os.environ["BILLING_AGENT_CARD_URL"],
                headers={"Authorization": f"Bearer {token}"} if token else None,
            )
        )

    if os.environ.get("HR_AGENT_CARD_URL"):
        token = os.environ.get("HR_AGENT_TOKEN")
        tools.append(
            define_local_a2a(
                name="intranet_hr_agent",
                agent_card_url=os.environ["HR_AGENT_CARD_URL"],
                headers={"Authorization": f"Bearer {token}"} if token else None,
            )
        )

    if not tools:
        print(
            "Set HR_AGENT_CARD_URL (and optionally BILLING_AGENT_CARD_URL) to a reachable "
            "Agent Card endpoint.",
            file=sys.stderr,
        )
        sys.exit(1)

    prompt = (
        sys.argv[1]
        if len(sys.argv) > 1
        else "When does the company holiday calendar reset for the new fiscal year?"
    )

    result = client.run_agent(
        system_prompt=(
            "You are a helpful router. Use `billing_agent` for billing questions "
            "and `intranet_hr_agent` for HR / time-off questions. If only one "
            "delegate is available, fall back to it. Reply with the delegate's answer."
        ),
        prompt=prompt,
        tools=tools,
        reasoning_level="medium",
        on_assistant_delta=lambda d: print(d, end="", flush=True),
    )
    print()
    print("---")
    print("Final reply:", result.text)


if __name__ == "__main__":
    main()

File: go/examples/a2a-tools/main.go
// Two flavours of Agent2Agent delegation in one ephemeral agent:
//
//   • mantyx.MantyxA2A — public peer MANTYX dials directly (server-resolved).
//   • mantyx.LocalA2A  — peer the SDK reaches on MANTYX's behalf
//     (client-resolved). Pass the Agent Card URL only — the SDK fetches
//     the card on the first run, ships it inline with the spec (so
//     MANTYX never reaches your intranet), and JSON-RPC `message/send`s
//     against the resolved card's `url` whenever MANTYX emits a
//     `local_tool_call` event for this tool.
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	mantyx "github.com/mantyx-io/mantyx-go-sdk"
)

func main() {
	apiKey := mustEnv("MANTYX_API_KEY")
	workspace := mustEnv("MANTYX_WORKSPACE_SLUG")

	opts := mantyx.Options{APIKey: apiKey, WorkspaceSlug: workspace}
	if base := os.Getenv("MANTYX_BASE_URL"); base != "" {
		opts.BaseURL = base
	}
	client := mantyx.NewClient(opts)

	tools := []mantyx.ToolRef{}

	if cardURL := os.Getenv("BILLING_AGENT_CARD_URL"); cardURL != "" {
		opts := mantyx.MantyxA2AOptions{
			Name:         "billing_agent",
			Description:  "Delegate billing questions to the public Acme billing agent.",
			AgentCardURL: cardURL,
		}
		if token := os.Getenv("BILLING_AGENT_TOKEN"); token != "" {
			opts.Headers = map[string]string{"Authorization": "Bearer " + token}
		}
		tools = append(tools, mantyx.MantyxA2A(opts))
	}

	if cardURL := os.Getenv("HR_AGENT_CARD_URL"); cardURL != "" {
		spec := mantyx.LocalA2ASpec{
			Name:         "intranet_hr_agent",
			AgentCardURL: cardURL,
		}
		if token := os.Getenv("HR_AGENT_TOKEN"); token != "" {
			spec.Headers = map[string]string{"Authorization": "Bearer " + token}
		}
		tools = append(tools, mantyx.LocalA2A(spec))
	}

	if len(tools) == 0 {
		fmt.Fprintln(os.Stderr, "Set HR_AGENT_CARD_URL (and optionally BILLING_AGENT_CARD_URL) to a reachable Agent Card endpoint.")
		os.Exit(1)
	}

	prompt := "When does the company holiday calendar reset for the new fiscal year?"
	if len(os.Args) > 1 {
		prompt = os.Args[1]
	}

	result, err := client.RunAgent(context.Background(), mantyx.RunSpec{
		SystemPrompt: "You are a helpful router. " +
			"Use `billing_agent` for billing questions and `intranet_hr_agent` for HR / time-off questions. " +
			"If only one delegate is available, fall back to it. Reply with the delegate's answer.",
		Prompt:         prompt,
		Tools:          tools,
		ReasoningLevel: mantyx.ReasoningMedium(),
		OnAssistantDelta: func(s string) {
			fmt.Print(s)
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println()
	fmt.Println("---")
	fmt.Println("Final reply:", result.Text)
}

func mustEnv(name string) string {
	v := os.Getenv(name)
	if v == "" {
		fmt.Fprintf(os.Stderr, "Missing required env var %s\n", name)
		os.Exit(1)
	}
	return v
}
